flang/lib/Lower/OpenMP/Utils.cpp

```diff
@@ -145,11 +145,294 @@ createMapInfoOp(fir::FirOpBuilder &builder, mlir::Location loc,
+      if (arrayExprWithSubscript(objectList[i])) {
+        llvm::SmallVector<mlir::Value> indices;
+        generateArrayIndices(converter, firOpBuilder, stmtCtx, clauseLocation,
+                             indices, objectList[i]);
+        assert(!indices.empty() && "missing expected indices for map clause");
+        curValue = firOpBuilder.create<fir::CoordinateOp>(
+            clauseLocation, firOpBuilder.getRefType(arrType.getEleTy()),
+            curValue, indices);
+      }
+    }
+
+    // If we encounter a record type, we must access the subsequent member
+    // by indexing into it and creating a coordinate operation to do so, we
+    // utilise the index information generated previously and passed in to
+    // work out the correct member to access and the corresponding member
+    // type.
+    if (fir::RecordType recordType = mlir::dyn_cast<fir::RecordType>(
+            fir::unwrapPassByRefType(curValue.getType()))) {
+      mlir::Value idxConst = firOpBuilder.createIntegerConstant(
+          clauseLocation, firOpBuilder.getIndexType(), indices[currentIndex]);
+      mlir::Type memberTy =
+          recordType.getTypeList().at(indices[currentIndex]).second;
+      curValue = firOpBuilder.create<fir::CoordinateOp>(
+          clauseLocation, firOpBuilder.getRefType(memberTy), curValue,
+          idxConst);
+
+      // Skip mapping and the subsequent load if we're the final member or not
+      // a type with a descriptor such as a pointer/allocatable. If we're a
+      // final member, the map will be generated by the processMap call that
+      // invoked this function, and if we're not a type with a descriptor then
+      // we have no need of generating an intermediate map for it, as we only
+      // need to generate a map if a member is a descriptor type (and thus
+      // obscures the members it contains via a pointer in which it's data needs
+      // mapped)
+      if ((currentIndex == indices.size() - 1) ||
+          !fir::isTypeWithDescriptor(memberTy)) {
+        currentIndex++;
+        continue;
+      }
+
+      llvm::SmallVector<int64_t> interimIndices(
+          indices.begin(), std::next(indices.begin(), currentIndex + 1));
+      // Verify we haven't already created a map for this particular member, by
+      // checking the list of members already mapped for the current parent,
+      // stored in the parentMemberIndices structure
+      if (!parentMemberIndices.isDuplicateMemberMapInfo(interimIndices)) {
+        // Generate bounds operations using the standard lowering utility,
+        // unfortunately this currently does a bit more than just generate
+        // bounds and we discard the other bits. May be useful to extend the
+        // utility to just provide bounds in the future.
+        llvm::SmallVector<mlir::Value> intermBounds;
+        if (i + 1 < objectList.size() &&
+            objectList[i + 1].sym()->IsObjectArray()) {
+          std::stringstream intermFortran;
```

Comment:
Oh I see, I only proposed that because I saw "interim" being used around as well. I don't mind how you name it. It looks like it's not used anywhere, just passed because it's a required argument, so maybe the name can reflect that.